Merge 2019-07-23-Intro to development

BohdanDudkaTD opened the pull request:
In 2019-07-23-Intro branch was created an vehicle classes hierarchy which contains both more abstract entities and concrete implementations. All classes are documented.
Also it contains tests for classes functionality.

dkalinin commented:
@BohdanDudkaTD don't forget about Pull request description.
Also let's have a hierarchy of folders according to inheritance.

dkalinin commented on 2019-07-23-Intro/tests/index.html:
```diff
@@ -0,0 +1,15 @@
+<div id="qunit"></div>
+<div id="qunit-fixture"></div>
+<script src="https://code.jquery.com/qunit/qunit-2.9.2.js"></script>
+<script type="module" src="tests.js"></script>
```
![image](https://user-images.githubusercontent.com/11214435/61789641-dc9be680-ae1d-11e9-8246-843aba629848.png)
Let's create `app` folder where all classes should be moved to.
Also, let's have a folder hierarchy in `app` which represents your class hierarchy.

BohdanDudkaTD replied:
All classes are moved to the `app` folder, now folders with classes are stored according to the inheritance hierarchy.
`tests` and `app` folders are in the root of `2019-07-23-Intro` folder.

dkalinin commented on 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js:
```diff
@@ -0,0 +1,40 @@
+     * @returns {boolean} true if passed in vehicle object can be carried by this object, otherwise - false.
+     */
+    canCarry(vehicle) {
+        return super.canCarry(vehicle);
```
This overriding is quite dummy. We don't need it.

BohdanDudkaTD replied:
Removed unnecessary overriding.

dkalinin commented on 2019-07-23-Intro/app/vehicle/aircraft/index.js:
```diff
@@ -0,0 +1,42 @@
+     * @param {number} volume - integer or floating number which describes volume of this aircraft in cubic meters.
+     * @param {string} movingEnvironment - environment in which this aircraft can move.
+     */
+    constructor(name, carryingVolume, carryingWeight, weight, volume, movingEnvironment) {
```
Why do we pass movingEnvironment to Aircraft constructor? Isn't it known?

BohdanDudkaTD replied:
Do you mean it should be a const field?

dkalinin commented on 2019-07-23-Intro/app/vehicle/aircraft/helicopter/index.js:
```diff
@@ -0,0 +1,40 @@
+     * can by carried by this helicopter.
+     * @param {number} weight - integer or floating number which describes weight of this helicopter in tonnes.
+     * @param {number} volume integer or floating number which describes volume of this helicopter in cubic meters.
+     * @param {string} movingEnvironment - environment in which this helicopter can move.
```
Let's have a separate class for 'movingEnvironment'.

BohdanDudkaTD replied:
Should it be some kind of enum?

BohdanDudkaTD commented:
@dkalinin, fixed, PTAL.

Fixed issues:
- extract possible moving environments for vehicles to separate class (MovingEnvironmentEnum).
Add relevant tests;
- also movingEnvironment is now not passed in constructor but assigned at field creation;
- remove unnecessary overriding;
- add unique fields for few classes;

BohdanDudkaTD commented:
@dkalinin fixed, PTAL.

Fixed issues:
- update JS docs, add override anotation;
- remove unnecessary fields from classes, move movementEnvironment field initialization to constructor;
- now Vehicle class can't be instantiated, get movingEnvironment can't be called directly;
- rename MovingEnvironmentEnum, remove unnecessary indefinite field;
- refactor code style, test messages;

BohdanDudkaTD commented:
Updates:

- add classes of predefined vehicles;
- add parameterized tests.

dkalinin commented on 2019-07-23-Intro/app/vehicle/bicycle/index.js:
```diff
@@ -0,0 +1,62 @@
+     * @returns {MovingEnvironment} enum`s {@link MovingEnvironment.GROUND} constant with string field.
+     */
+    get movingEnvironment() {
+        return this._movingEnvironment;
```
There's no need to have a private property `_movingEnvironment ` just return `MovingEnvironment.GROUND`.

dkalinin commented on 2019-07-23-Intro/tests/fishBoatTests.js:
```diff
@@ -0,0 +1,40 @@
+            {
+                vehicleInstance: new CannondaleCatalyst2("Speed"),
+                expectedCanCarryResult: true,
+                assertionMessage: "Fish boat should carry bicycle."
```
Cannondale is a bicycle. Check other assertionMessages too.

BohdanDudkaTD replied:
Could you, please, explain more detailed where exactly is the mistake?
Here a fish boat is the testable object and a bicycle is one of vehicle`s instances which should be carried by a fish boat.

dkalinin commented on 2019-07-23-Intro/tests/fishBoatTests.js:
```diff
@@ -0,0 +1,40 @@
+});
+QUnit.test("Test fish boat`s canCarry.", function (assert) {
+    const fishBoat = new BaylinerElementF16("Fishing");
+    this.testObject.forEach(element => {
```
It's better to create create `QUnit.test`in function which iterates testObject.
